### PR TITLE
Add Team Usage

### PR DESCRIPTION
**Pull request by evisos-ons**
### Main Changes
Created additional page. Main page is now viewable in org_usage and new team_usage available now.

### Flow
User logs in with GitHub, views their teams, selects a team and view the team CoPilot usage data. If the user is part of "keh-dev" they can search for other teams to view their data.

### Notes
UI is replicated from existing UI, with addition of delta to the metrics.
View how to setup the dashboard in the README file.

### CODE REVIEW

File: README.md
Comment (gibbardsteve) [inline README.md:1]: Can you add a comment for the makefile options (once they are clarified between yourself and Hasan)

File: README.md [hunk 1]
```diff
@@ -1,3 +1,44 @@
+# add_team_usage_auth
+
+For testing purposes, create a 'New OAuth App'. Go to this [page](https://github.com/settings/developers) where you can create the app. Note down the Client ID and Client secret.
+
+- Set the application name to anything.
```
Comment (gibbardsteve) [inline README.md:5]: It is worth giving a sensical example - e.g github-auth-app

File: src/app.py [hunk 2]
```diff
@@ -1,637 +1,12 @@
...
-# Initialise a boto3 session
-session = boto3.Session()
 
 st.set_page_config(
     page_title="CoPilot Usage Dashboard",
     page_icon="./src/branding/ONS-symbol_digital.svg",
     layout="wide",
 )
+create_page = st.Page("./pages/org_usage.py", title="Organization Usage", icon=":material/groups:")
+delete_page = st.Page("./pages/team_usage.py", title="Team Usage", icon=":material/group:")
```
Comment (gibbardsteve) [inline src/app.py:9]: Why are the variable names create_page and delete_page ?  Couldn't it be org_page and team_page?

File: src/auth.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+org_client_id = os.getenv("GITHUB_APP_CLIENT_ID")
+
+# AWS Secret Manager Secret Name for the .pem file
+secret_name = os.getenv("AWS_SECRET_NAME")
+secret_reigon = os.getenv("AWS_DEFAULT_REGION")
+
+account = os.getenv("AWS_ACCOUNT_NAME")
+
+
+client_id = "Ov23liJPFTZTiHwrQiEq"
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply src/auth.py:22]: This was removed and I added a comment there to explain that.

File: src/pages/team_usage.py [hunk 4]
```diff
@@ -0,0 +1,664 @@
...
+
+        except Exception as e:
+            # This would be an error with either getting the access token or getting their profile
+            st.error(f"Error during login: {e}")
+            st.stop()
+    else:
+        # Login button. User is directed to GitHub oauth page then once authorized they come back to this page and go to the next step
+        login_url = f"{authorize_url}?{urlencode({'client_id': client_id, 'redirect_uri': redirect_uri, 'scope': 'user:email read:org'})}"
+        st.markdown(
+            f'<a href="{login_url}" target="_self">:blue-background[Login with GitHub]</a>', unsafe_allow_html=True
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply src/pages/team_usage.py:269]: Putting this link in a button opens the page in a new tab, putting the link in markdown with `unsafe_allow_html` allows the link to be opened in the same tab. As each page refresh clears the session, for security reasons, the user has to keep logging in. I added this so the user doesn't end up having an overload of tabs open.

I can remove this and make it back into a button if you think that would be better, but this `unsafe_allow_html` is safe.

File: README.md [hunk 5]
```diff
@@ -1,3 +1,44 @@
+# add_team_usage_auth
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply README.md:1]: This was a temporary section but I'll change that now.

File: README.md (hunk 1 above)
Comment (evisos-ons) [reply README.md:5]: Added this - although apps can't start with github or git.

File: aws_lambda_scripts/team.py
[the remark this answers is not in the document]
Comment (evisos-ons) [reply aws_lambda_scripts/team.py:1]: Will remove that from the PR.

File: src/app.py (hunk 2 above)
Comment (evisos-ons) [reply src/app.py:9]: Changed.

File: src/auth.py
[the remark this answers is not in the document]
Comment (evisos-ons) [reply src/auth.py:1]: Removed and secrets have been refreshed so they're unusable.

File: src/pages/team_usage.py [hunk 6]
```diff
@@ -0,0 +1,664 @@
...
+    # If user is in the org, get the access token. If not display an error and stop
+    if not is_user_in_org(st.session_state.profile["login"], org):
+        st.error(f"Sorry, {st.session_state.profile['login']}, you are not part of the {org} organization.")
+        st.stop()
+
+    # Get the users teams
+    user_teams = get_user_teams(access_token[0], st.session_state.profile)
+
+    if access_token and user_teams:
+        if "keh-dev" in user_teams:
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply src/pages/team_usage.py]: Added  admin_teams.json file to copilot-usage-dashboard bucket. Admin can add or remove teams from the list. At the moment the list looks like this:

`["keh-dev", "sdp-dev"]`

File: src/pages/team_usage.py [hunk 7]
```diff
@@ -0,0 +1,664 @@
...
+            input_method = st.radio(
+                "You are part of an admin team, so you can either:",
+                ("Select your team", "Search for a team"),
+                horizontal=True,
+            )
+
+            if input_method == "Select your team":
+                team_slug = st.selectbox("Select team:", options=user_teams)
+            else:
+                team_slug = st.text_input("Enter team name:", value="keh-dev")
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply src/pages/team_usage.py:322]: Could you elaborate on this please?

File: Makefile [hunk 8]
```diff
@@ -19,18 +16,22 @@ format:  ## Format the code.
...
+	poetry run black aws_lambda_scripts --check
 
 .PHONY: ruff
-ruff:
+ruff: ## Run ruff without fixing.
+	poetry run ruff check src
+	poetry run ruff check aws_lambda_scripts
+
+.PHONY: ruff-fix
+ruff-fx: ## Run ruff with fixing.
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply Makefile]: Removed this.

File: README.md
Comment (evisos-ons) [reply README.md:1]: Added and moved.

File: aws_lambda_scripts/team.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+access_token = github_api_toolkit.get_token_as_installation(org, secret, client_id)
+print(access_token)
+
+# Create an instance of the api_controller class
+gh = github_api_toolkit.github_interface(access_token[0])
+
+
+# Get the usage data with query parameters
+usage_data = gh.get(
+    f"/orgs/{org}/team/keh-dev/copilot/usage",
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply aws_lambda_scripts/team.py]: Removed.

File: src/pages/team_usage.py [hunk 10]
```diff
@@ -0,0 +1,664 @@
...
+# Streamlit UI starts here
+st.logo("./src/branding/ONS_Logo_Digital_Colour_Landscape_Bilingual_RGB.svg")
+col1, col2 = st.columns([0.8, 0.2])
+
+# Header
+col1.title(":blue-background[GitHub Team Copilot Usage]")
+
+col2.image("./src/branding/ONS_Logo_Digital_Colour_Landscape_Bilingual_RGB.png")
+
+# Initialize session states
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply src/pages/team_usage.py]: Added calculate_deltas()

File: src/pages/team_usage.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (evisos-ons) [reply src/pages/team_usage.py]: Changed to user_teams[0] or nothing.

File: src/pages/team_usage.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (evisos-ons) [reply src/pages/team_usage.py]: Used st.html instead, thanks!

File: README.md [hunk 11]
```diff
@@ -62,6 +62,46 @@ export AWS_ACCOUNT_NAME=sdp-sandbox
...
+```bash
+export CLIENT_ID=<client_id>
+export CLIENT_SECRET=<client_secret>
+```
+
+Make sure no other apps are running on `localhost:8502`. Once setup, use `make run-local`. This runs the the streamlit app on port **8502**, as *8501* is the default port.
+
+Once the app is running, head to `localhost:8502/team_usage` or by clicking the `Team Usage` in the sidebar. Click the `Login with GitHub` button. This redirects the user to the GitHub OAuth page. Click the green `Authorize` button, which redirects you back to the main application and logs you in.
+
+If you are part of the `keh-dev` team then you can either select a team that you are in from the select box or you can enter another team name. As of 30-09-2024, there are only these teams with team copilot data: `all`, `Blaise5`, `CSS`, `keh-dev`, `Ops`.
```
[the remark this answers is not in the document]
Comment (evisos-ons) [reply README.md:96]: Added explanation.

**Review by gibbardsteve — COMMENTED** [review]
Looks good, it looks like you need to merge some changes

**Comment by evisos-ons** [comment]
> Looks good, it looks like you need to merge some changes

Resolved conflicts, ready for merge.

File: CODEOWNERS [hunk 12]
```diff
@@ -1,2 +1,3 @@
 @TotalDwarf03
+@evisos-ons
```
Comment (gibbardsteve) [inline CODEOWNERS:2]: Technically I don't think you should need this (likewise Kieran's username) as you are in the keh-dev team that is mentioned, happy for it to stay though.

**Review by gibbardsteve — APPROVED** [review]

**Pull request merged**